rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala

```diff
@@ -63,4 +72,98 @@ class CostAccountingSpec extends FlatSpec with TripleEqualsSupport {
+                                _ <- runtime.close()
+                                _ <- Task.eval(dbDir.recursivelyDelete())
+                              } yield (res)
+                            }.attempt)
+      (result, costLog) = costsLoggingProgram
+      res               <- errorLog.readAndClearErrorVector
+      _                 = Task.now(res should be(Vector.empty))
+    } yield ((result, costLog))).unsafeRunSync
+  }
+
+  "Total cost of evaluation" should "be equal to the sum of all costs in the log" ignore {
+    val initialPhlo = 10000L
+    val contract    = """new loop in {
+                             contract loop(@n) = {
+                               match n {
+                                 0 => Nil
+                                 _ => loop!(n-1)
+                               }
+                             } |
+                             loop!(10)
+                           }""".stripMargin
+
+    val expectedTotalCost = 1766L
+    val (result, costLog) = evaluateWithCostLog(initialPhlo, contract)
+    result shouldBe Right(EvaluateResult(Cost(expectedTotalCost), Vector.empty))
+    costLog.map(_.value).toList.sum shouldEqual expectedTotalCost
+  }
+
+  "Running out of phlogistons" should "stop the evaluation upon cost depletion in a single execution branch" in {
+    val initialPhlo       = 1L
+    val contract          = "@1!(1)"
+    val expectedCosts     = List(Cost(4, "substitution"))
+    val (result, costLog) = evaluateWithCostLog(initialPhlo, contract)
+    result shouldBe Left(OutOfPhlogistonsError)
+    costLog.toList should contain theSameElementsAs (expectedCosts)
+  }
+
+  it should "stop the evaluation of all execution branches when one of them runs out of phlo" ignore {
+    val initialPhlo       = 1L
+    val contract          = "@1!(1) | @2!(2) | @3!(3)"
+    val expectedCosts     = List(Cost(4, "substitution"))
+    val (result, costLog) = evaluateWithCostLog(initialPhlo, contract)
+    result shouldBe Left(OutOfPhlogistonsError)
+    costLog.toList should contain theSameElementsAs (expectedCosts)
+  }
+
+  it should "stop the evaluation of all execution branches when one of them runs out of phlo with a more sophisiticated contract" ignore {
+    val initialPhlo = 150L
+    val contract    = """new loop in {
+                             contract loop(@n) = {
+                               match n {
+                                 0 => Nil
+                                 _ => loop!(n-1)
+                               }
+                             } |
+                             loop!(10)
+                           }""".stripMargin
+
+    val (result, costLog) = evaluateWithCostLog(initialPhlo, contract)
+    result shouldBe Left(OutOfPhlogistonsError)
+    val costs = costLog.map(_.value).toList
+    // The sum of all costs but last needs to be <= initialPhlo, otherwise
+    // the last cost should have not been logged
+    costs.init.sum.toLong should be <= (initialPhlo)
```

Comment:
So, basically, by checking that the total cost of all the operations except the last are less than the phlo limit, we can ensure that the last operation was the operation to exceed the phlo limit. Since no more costs were logged after it, we know that all execution branches were terminated?